thriftbp/thrifttest/server.go

```diff
@@ -0,0 +1,211 @@
+// waits for a short period of time to give the server time to start up.
+//
+// The server can be shut down manually using server.Close, with the shutdown
+// commands defined in runtimebp, or by cancelling the given context.
+func (s *Server) Start(ctx context.Context) {
+	go baseplate.Serve(ctx, s)
+	time.Sleep(10 * time.Millisecond)
+}
+
+// Close both the underying baseplate.Server and thriftbp.ClientPool
+func (s *Server) Close() error {
+	var errs batcherror.BatchError
+	if err := s.Server.Close(); err != nil {
+		errs.Add(err)
+	}
+	if s.ClientPool != nil {
+		if err := s.ClientPool.Close(); err != nil {
+			errs.Add(err)
+		}
+	}
+	return errs.Compile()
+}
+
+// NewBaseplateServer returns a new, Baseplate thrift server listening on the
+// local loopback interface and a Baseplate ClientPool for use with that server.
+//
+// This is inspired by httptest.NewServer from the go standard library and can
+// be used to test a thrift service.
+//
+// "cfg" may be nil, if it is, sane defaults will be chosen.
+// The server and pool that are returned should be closed when done, but the
+// Baseplate used by the server does not need to be.
+//
+// Here is an example usage of NewBaseplateServer:
+//
+//	import (
+//		"context"
+//		"testing"
+//
+//		"github.com/reddit/baseplate.go/batcherror"
+//		baseplatethrift "github.com/reddit/baseplate.go/internal/gen-go/reddit/baseplate"
+//		"github.com/reddit/baseplate.go/secrets"
+//		"github.com/reddit/baseplate.go/thriftbp/thrifttest"
+//	)
+//
+//	type BaseplateService struct {
+//		Fail bool
+//		Err  error
+//	}
+//
+//	func (srv BaseplateService) IsHealthy(ctx context.Context) (r bool, err error) {
+//		return !srv.Fail, srv.Err
+//	}
+//
+//	func TestService(t *testing.T){
+//		// Initialize this properly in a real test
+//		var secrets *secrets.Store
+//
+//		ctx, cancel := context.WithCancel(context.Background())
+//		defer cancel()
+//
+//		processor := baseplatethrift.NewBaseplateServiceProcessor(BaseplateService{})
+//		server, err := thrifttest.NewBaseplateServer(store, processor, nil)
+//		if err != nil {
+//			t.Fatal(err)
+//		}
+//		// cancelling the context will close the server.
+//		server.Start(ctx)
+//
+//		client := baseplatethrift.NewBaseplateServiceClient(server.ClientPool)
+//		success, err := client.IsHealthy(ctx)
+//
+//		if err != nil {
+//			t.Errorf("expected no error, got %v", err)
+//		}
+//
+//		if !success {
+// 			t.Errorf("result mismatch, expected %v, got %v", c.expected.result, result)
+//		}
+//	}
+func NewBaseplateServer(
+	store *secrets.Store,
+	processor thrift.TProcessor,
+	cfg *ServerConfig,
+) (*Server, error) {
+	if cfg == nil {
+		cfg = &ServerConfig{}
+	}
+
+	socket, err := thrift.NewTServerSocket(loopbackAddr)
+	if err != nil {
+		return nil, err
+	}
+	// Call listen to reserve a port and check for any issues early
+	if err := socket.Listen(); err != nil {
+		return nil, err
+	}
+
+	cfg.ServerConfig.Addr = socket.Addr().String()
```

Comment:
one side effect of using pointer for `cfg` is that if the passed in cfg is non-nil, then these kind of changes will be reflected back to the caller (e.g. the values in `cfg` will change after calling `NewBaseplateServer`), which in most cases won't be an issue but it's something to be mindful about.